File: basler_pyqt6/ui/widgets/camera_control.py
"""
相機控制組件
"""

from PyQt6.QtWidgets import (
    QWidget, QVBoxLayout, QHBoxLayout, QGroupBox,
    QPushButton, QListWidget, QLabel, QSlider, QListWidgetItem
)
from PyQt6.QtCore import Qt, pyqtSignal


class CameraControlWidget(QWidget):
    """相機控制組件"""

    # 信號
    detect_clicked = pyqtSignal()
    connect_clicked = pyqtSignal(int)
    disconnect_clicked = pyqtSignal()
    start_clicked = pyqtSignal()
    stop_clicked = pyqtSignal()
    exposure_changed = pyqtSignal(float)

    def __init__(self, parent=None):
        super().__init__(parent)
        self.selected_camera_index = None
        self.init_ui()

    def init_ui(self):
        """初始化 UI"""
        layout = QVBoxLayout(self)

        # 創建分組框
        group_box = QGroupBox("📷 相機控制")
        group_layout = QVBoxLayout()

        # 檢測按鈕
        self.detect_btn = QPushButton("🔍 檢測相機")
        self.detect_btn.clicked.connect(self.detect_clicked.emit)
        group_layout.addWidget(self.detect_btn)

        # 相機列表
        self.camera_list = QListWidget()
        self.camera_list.setMaximumHeight(150)
        self.camera_list.itemClicked.connect(self.on_camera_selected)
        group_layout.addWidget(QLabel("可用相機:"))
        group_layout.addWidget(self.camera_list)

        # 連接/斷開按鈕
        btn_layout = QHBoxLayout()
        self.connect_btn = QPushButton("連接")
        self.connect_btn.setEnabled(False)
        self.connect_btn.clicked.connect(self.on_connect_clicked)

        self.disconnect_btn = QPushButton("斷開")
        self.disconnect_btn.setEnabled(False)
        self.disconnect_btn.clicked.connect(self.disconnect_clicked.emit)

        btn_layout.addWidget(self.connect_btn)
        btn_layout.addWidget(self.disconnect_btn)
        group_layout.addLayout(btn_layout)

        # 開始/停止抓取
        grab_layout = QHBoxLayout()
        self.start_btn = QPushButton("▶️ 開始抓取")
        self.start_btn.setEnabled(False)
        self.start_btn.clicked.connect(self.start_clicked.emit)

        self.stop_btn = QPushButton("⏸️ 停止抓取")
        self.stop_btn.setEnabled(False)
        self.stop_btn.clicked.connect(self.stop_clicked.emit)

        grab_layout.addWidget(self.start_btn)
        grab_layout.addWidget(self.stop_btn)
        group_layout.addLayout(grab_layout)

        # 曝光控制
        group_layout.addWidget(QLabel("曝光時間 (μs):"))
        exposure_layout = QHBoxLayout()

        self.exposure_slider = QSlider(Qt.Orientation.Horizontal)
        self.exposure_slider.setMinimum(100)
        self.exposure_slider.setMaximum(10000)
        self.exposure_slider.setValue(1000)
        self.exposure_slider.setEnabled(False)
        self.exposure_slider.valueChanged.connect(self.on_exposure_changed)

        self.exposure_label = QLabel("1000")
        exposure_layout.addWidget(self.exposure_slider)
        exposure_layout.addWidget(self.exposure_label)
        group_layout.addLayout(exposure_layout)

        group_box.setLayout(group_layout)
        layout.addWidget(group_box)

    def update_camera_list(self, cameras):
        """更新相機列表"""
        self.camera_list.clear()

        for camera in cameras:
            item_text = f"{camera['model']} ({camera['serial']})"
            if camera.get('is_target', False):
                item_text += " ✓"

            item = QListWidgetItem(item_text)
            item.setData(Qt.ItemDataRole.UserRole, camera['index'])
            self.camera_list.addItem(item)

    def on_camera_selected(self, item):
        """相機選中"""
        self.selected_camera_index = item.data(Qt.ItemDataRole.UserRole)
        self.connect_btn.setEnabled(True)

    def on_connect_clicked(self):
        """連接按鈕點擊"""
        if self.selected_camera_index is not None:
            self.connect_clicked.emit(self.selected_camera_index)
            self.disconnect_btn.setEnabled(True)
            self.start_btn.setEnabled(True)
            self.exposure_slider.setEnabled(True)

    def on_exposure_changed(self, value):
        """曝光值改變"""
        self.exposure_label.setText(str(value))
        self.exposure_changed.emit(float(value))

    def set_grabbing_state(self, grabbing):
        """設置抓取狀態"""
        if grabbing:
            self.start_btn.setEnabled(False)
            self.stop_btn.setEnabled(True)
        else:
            self.start_btn.setEnabled(True)
            self.stop_btn.setEnabled(False)

    def set_video_mode(self, is_video: bool):
        """設置為視頻模式"""
        if is_video:
            # 視頻模式：隱藏相機相關控件
            self.detect_btn.setEnabled(False)
            self.camera_list.setEnabled(False)
            self.connect_btn.setEnabled(False)
            self.disconnect_btn.setEnabled(False)
            self.exposure_slider.setEnabled(False)
            self.start_btn.setEnabled(True)
            self.start_btn.setText("▶️ 播放")
            self.stop_btn.setText("⏸️ 暫停")
        else:
            # 相機模式：恢復相機控件
            self.detect_btn.setEnabled(True)
            self.camera_list.setEnabled(True)
            self.start_btn.setText("▶️ 開始抓取")
            self.stop_btn.setText("⏸️ 停止抓取")

File: basler_cpp/src/ui/widgets/camera_control.cpp
#include "ui/widgets/camera_control.h"
#include <QVBoxLayout>
#include <QHBoxLayout>

namespace basler {

CameraControlWidget::CameraControlWidget(QWidget* parent)
    : QWidget(parent)
{
    initUi();
}

void CameraControlWidget::initUi()
{
    QVBoxLayout* mainLayout = new QVBoxLayout(this);

    m_groupBox = new QGroupBox(tr("📷 相機控制"));
    QVBoxLayout* groupLayout = new QVBoxLayout();

    // Detection buttons layout
    QHBoxLayout* detectLayout = new QHBoxLayout();

    // Quick detect button (single scan)
    m_detectBtn = new QPushButton(tr("Quick Scan"));
    m_detectBtn->setToolTip(tr("Single scan - fast but may miss cameras that are booting"));
    connect(m_detectBtn, &QPushButton::clicked, this, &CameraControlWidget::detectRequested);
    detectLayout->addWidget(m_detectBtn);

    // Auto-retry detect button (recommended)
    m_autoDetectBtn = new QPushButton(tr("🔍 Auto-Detect (Recommended)"));
    m_autoDetectBtn->setToolTip(tr("Smart scan with 3 auto-retries - finds cameras that need time to boot"));
    m_autoDetectBtn->setStyleSheet("font-weight: bold;");
    connect(m_autoDetectBtn, &QPushButton::clicked, this, &CameraControlWidget::detectWithRetryRequested);
    detectLayout->addWidget(m_autoDetectBtn);

    groupLayout->addLayout(detectLayout);

    // 相機選擇（ComboBox）
    m_cameraCombo = new QComboBox();
    m_cameraCombo->setPlaceholderText(tr("選擇相機..."));
    groupLayout->addWidget(m_cameraCombo);

    // 連接/斷開按鈕
    QHBoxLayout* connectLayout = new QHBoxLayout();
    m_connectBtn = new QPushButton(tr("連接"));
    m_connectBtn->setEnabled(false);
    connect(m_connectBtn, &QPushButton::clicked, this, &CameraControlWidget::connectRequested);
    connectLayout->addWidget(m_connectBtn);

    m_disconnectBtn = new QPushButton(tr("斷開"));
    m_disconnectBtn->setEnabled(false);
    connect(m_disconnectBtn, &QPushButton::clicked, this, &CameraControlWidget::disconnectRequested);
    connectLayout->addWidget(m_disconnectBtn);
    groupLayout->addLayout(connectLayout);

    // 抓取控制按鈕
    QHBoxLayout* grabLayout = new QHBoxLayout();
    m_startBtn = new QPushButton(tr("▶ 開始"));
    m_startBtn->setEnabled(false);
    connect(m_startBtn, &QPushButton::clicked, this, &CameraControlWidget::startGrabRequested);
    grabLayout->addWidget(m_startBtn);

    m_stopBtn = new QPushButton(tr("⏹ 停止"));
    m_stopBtn->setEnabled(false);
    connect(m_stopBtn, &QPushButton::clicked, this, &CameraControlWidget::stopGrabRequested);
    grabLayout->addWidget(m_stopBtn);
    groupLayout->addLayout(grabLayout);

    // 曝光控制
    QHBoxLayout* exposureLayout = new QHBoxLayout();
    exposureLayout->addWidget(new QLabel(tr("曝光:")));

    m_exposureSlider = new QSlider(Qt::Horizontal);
    m_exposureSlider->setRange(100, 100000);  // 100us - 100ms
    m_exposureSlider->setValue(10000);        // 默認 10ms
    connect(m_exposureSlider, &QSlider::valueChanged,
            this, &CameraControlWidget::onExposureChanged);
    exposureLayout->addWidget(m_exposureSlider);

    m_exposureLabel = new QLabel("10000 μs");
    m_exposureLabel->setMinimumWidth(80);
    exposureLayout->addWidget(m_exposureLabel);
    groupLayout->addLayout(exposureLayout);

    m_groupBox->setLayout(groupLayout);
    mainLayout->addWidget(m_groupBox);
    mainLayout->addStretch();

    // 連接 combo box 選擇變更
    connect(m_cameraCombo, QOverload<int>::of(&QComboBox::currentIndexChanged),
            [this](int index) {
                m_connectBtn->setEnabled(index >= 0 && !m_isConnected && !m_isVideoMode);
            });
}

void CameraControlWidget::setCameraList(const QStringList& cameras)
{
    m_cameraCombo->clear();
    m_cameraCombo->addItems(cameras);

    if (!cameras.isEmpty()) {
        m_cameraCombo->setCurrentIndex(0);
        m_connectBtn->setEnabled(!m_isConnected && !m_isVideoMode);
    } else {
        m_connectBtn->setEnabled(false);
    }
}

void CameraControlWidget::setConnected(bool connected)
{
    m_isConnected = connected;
    updateButtonStates();
}

void CameraControlWidget::setGrabbing(bool grabbing)
{
    m_isGrabbing = grabbing;
    updateButtonStates();
}

void CameraControlWidget::setVideoMode(bool isVideo)
{
    m_isVideoMode = isVideo;

    // 視頻模式下禁用相機相關控制
    m_detectBtn->setEnabled(!isVideo);
    m_cameraCombo->setEnabled(!isVideo);
    m_connectBtn->setEnabled(!isVideo && m_cameraCombo->count() > 0 && !m_isConnected);
    m_disconnectBtn->setEnabled(!isVideo && m_isConnected);
    m_exposureSlider->setEnabled(!isVideo);

    if (isVideo) {
        m_startBtn->setEnabled(true);
        m_stopBtn->setEnabled(false);
    }
}

void CameraControlWidget::updateButtonStates()
{
    if (m_isVideoMode) return;

    m_detectBtn->setEnabled(!m_isConnected);
    m_cameraCombo->setEnabled(!m_isConnected);
    m_connectBtn->setEnabled(!m_isConnected && m_cameraCombo->count() > 0);
    m_disconnectBtn->setEnabled(m_isConnected && !m_isGrabbing);
    m_startBtn->setEnabled(m_isConnected && !m_isGrabbing);
    m_stopBtn->setEnabled(m_isGrabbing);
    m_exposureSlider->setEnabled(m_isConnected && !m_isGrabbing);
}

void CameraControlWidget::onExposureChanged(int value)
{
    m_exposureLabel->setText(QString("%1 μs").arg(value));
    emit exposureChanged(static_cast<double>(value));
}

} // namespace basler
